File: 1705-maximum-number-of-eaten-apples/1705-maximum-number-of-eaten-apples.cpp
class Solution {
public:
    int eatenApples(vector<int>& apples, vector<int>& days) {
	int eaten = 0, day = 0;
	priority_queue<pair<int, int>, vector<pair<int, int>>, greater<pair<int, int>>> pq;
	while (day < days.size() || !pq.empty()) {
		if (day < days.size() && apples[day] > 0) {
			pq.push( { day + days[day], apples[day] } );
		}
		while (!pq.empty()) {
			auto cur = pq.top();
			pq.pop();
			if (cur.second == 0 || cur.first <= day)
				continue;
			eaten++;
			if (cur.second > 1)
				pq.push( { cur.first, cur.second - 1 } );
			break;
		}
		day++;
	}
	return eaten;
}
};

File: 1705-maximum-number-of-eaten-apples/1705-maximum-number-of-eaten-apples.java
class Solution {
    public int eatenApples(int[] apples, int[] days) {
        PriorityQueue<int[]> appleStorage = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        int appleCount = 0;
        
        for(int i = 0; i < apples.length || !appleStorage.isEmpty(); i++) {
            if(i < apples.length) {
                appleStorage.add(new int[]{ i + days[i], apples[i] });
            }
            
            while(!appleStorage.isEmpty() && (appleStorage.peek()[0] <= i || appleStorage.peek()[1] <= 0)) {
                appleStorage.poll();
            }
            
            if(!appleStorage.isEmpty()) {
                appleCount++;
                appleStorage.peek()[1]--;
            }
        }
        
        return appleCount;
    }
}